### PR TITLE
Invert privVal socket communication

### PR DESCRIPTION
**Pull request by xla**
Follow-up to #1255 aligning with the expectation that the external signing process connects to the node. The SocketClient will block on start until one connection has been established, support for multiple signers connected simultaneously is a planned future extension.

* SocketClient accepts connection
* PrivValSocketServer renamed to RemoteSigner
* extend tests

### CODE REVIEW

File: types/priv_validator/socket.go [hunk 1]
```diff
@@ -62,24 +74,46 @@ func NewSocketClient(
...
 
-	sc.BaseService = *cmn.NewBaseService(logger, "privValidatorSocketClient", sc)
+	sc.BaseService = *cmn.NewBaseService(logger, "SocketClient", sc)
 
 	return sc
 }
 
 // OnStart implements cmn.Service.
 func (sc *SocketClient) OnStart() error {
 	if err := sc.BaseService.OnStart(); err != nil {
```
Comment (melekes) [inline types/priv_validator/socket.go]: BaseService.OnStart is empty and can be removed

File: types/priv_validator/socket.go [hunk 2]
```diff
@@ -195,166 +249,211 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
...
 
-func (pvss *PrivValidatorSocketServer) acceptConnections() {
-	for {
-		conn, err := pvss.listener.Accept()
+func (rs *RemoteSigner) connect() (net.Conn, error) {
+	retries := defaultDialRetries
+
+RETRY_LOOP:
+	for retries > 0 {
+		if retries != defaultDialRetries {
```
Comment (melekes) [inline types/priv_validator/socket.go]: I'd add a comment like `sleep if it's not our first retry`

File: types/priv_validator/socket.go [hunk 3]
```diff
@@ -62,24 +74,37 @@ func NewSocketClient(
...
+	sc.BaseService = *cmn.NewBaseService(logger, "SocketClient", sc)
 
 	return sc
 }
 
 // OnStart implements cmn.Service.
 func (sc *SocketClient) OnStart() error {
-	if err := sc.BaseService.OnStart(); err != nil {
-		return err
+	if sc.listener == nil {
```
Comment (ebuchman) [inline types/priv_validator/socket.go]: dont think we need this check.

File: types/priv_validator/socket.go [hunk 4]
```diff
@@ -195,166 +240,206 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
...
-				"sc connect",
-				"addr", sc.addr,
-				"err", errors.Wrap(err, "connection failed"),
-			)
+	if err := conn.SetDeadline(time.Now().Add(sc.connDeadline)); err != nil {
+		return nil, err
+	}
 
-			continue RETRY_LOOP
+	if sc.privKey != nil {
```
Comment (ebuchman) [inline types/priv_validator/socket.go]: lets require the privKey always :)

File: types/priv_validator/socket.go [hunk 5]
```diff
@@ -195,166 +240,206 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
...
+	)
+
+	go func(connc chan<- net.Conn, errc chan<- error) {
+		conn, err := sc.acceptConnection()
+		if err != nil {
+			errc <- err
+			return
 		}
 
+		connc <- conn
```
Comment (ebuchman) [inline types/priv_validator/socket.go]: should we include a way to exit this routine if the timeout triggerS ?

File: types/priv_validator/socket.go [hunk 6]
```diff
@@ -195,166 +240,206 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
...
+	if err := rs.conn.Close(); err != nil {
+		rs.Logger.Error("OnStop", "err", errors.Wrap(err, "closing listener failed"))
 	}
 }
 
-func (pvss *PrivValidatorSocketServer) acceptConnections() {
-	for {
-		conn, err := pvss.listener.Accept()
+func (rs *RemoteSigner) connect() (net.Conn, error) {
+	retries := defaultDialRetries
```
Comment (ebuchman) [inline types/priv_validator/socket.go]: use rs.connRetries

File: types/priv_validator/socket.go [hunk 7]
```diff
@@ -195,166 +240,206 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
...
+
+RETRY_LOOP:
+	for retries > 0 {
```
Comment (ebuchman) [inline types/priv_validator/socket.go]: lets put the `retries := ` and the `retries --` in the one liner

File: types/priv_validator/socket.go [hunk 8]
```diff
@@ -195,166 +240,206 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
...
+		// Don't sleep if it is the first retry.
+		if retries != defaultDialRetries {
```
Comment (ebuchman) [inline types/priv_validator/socket.go]: rs.connRetries

File: types/priv_validator/socket.go [hunk 9]
```diff
@@ -195,166 +240,206 @@ func (sc *SocketClient) SignHeartbeat(chainID string, heartbeat *types.Heartbeat
...
+			rs.Logger.Error(
+				"connect",
 				"err", errors.Wrap(err, "setting connection timeout failed"),
 			)
 			continue
 		}
 
-		if pvss.privKey != nil {
-			conn, err = p2pconn.MakeSecretConnection(conn, pvss.privKey.Wrap())
+		if rs.privKey != nil {
```
Comment (ebuchman) [inline types/priv_validator/socket.go]: require it

Comment (ebuchman) [reply types/priv_validator/socket.go]: we also probably want to authenticate. ie. we should be dialing a `ID@host:port` and then after the secret connection is setup, check that the ID (ie. hash of the remote pubkey) of the connection matches the ID we dialed (see how p2p does it)

**Pull request merged**